### PR TITLE
Update test project to net5 and CodeAnalysis 3.8.0-5.final

### LINKED ISSUE #3673: Fix unit tests for .NET 5 execution
Several steps:

- [x] - verify if  we can update ` <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.Common" Version="3.4.0" />` to parse and compile .net 5 code

- [x] - fix the failing test cases

### PR DESCRIPTION
**Pull request by andrei-epure-sonarsource**
Fixes #3673

### CODE REVIEW

**Comment by andrei-epure-sonarsource** [comment]
locally, I have one failing test
```
  Failed RestrictDeserializedTypesFormatters [68 ms]
  Error Message:
   Test method SonarAnalyzer.UnitTest.Rules.RestrictDeserializedTypesTest.RestrictDeserializedTypesFormatters threw exception:
SonarAnalyzer.UnitTest.TestFramework.UnexpectedDiagnosticException: CSharp8: Unexpected primary issue on line 187, span (186,23)-(186,34) with message 'Restrict types of objects allowed to be deserialized.'.
See output to see all actual diagnostics raised on the file
```

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataReference.cs [hunk 1]
```diff
@@ -31,16 +31,16 @@ internal static class NuGetMetadataReference
         public static References MicrosoftVisualStudioQualityToolsUnitTestFramework => Create("VS.QualityTools.UnitTestFramework", "15.0.27323.2");
         public static References MSTestTestFrameworkV1 => Create("MSTest.TestFramework", "1.1.11");
         public static References XunitFrameworkV1 =>
-            Create("xunit", "1.9.1")
-            .Concat(Create("xunit.extensions", "1.9.1"));
+            CreateWithCommandLine("xunit", "1.9.1")
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataReference.cs:34]: see https://github.com/SonarSource/sonar-dotnet/issues/3674

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/CallerInformationParametersShouldBeLastTest.cs [hunk 2]
```diff
@@ -40,7 +40,8 @@ public void CallerInformationParametersShouldBeLast()
         [TestCategory("Rule")]
         public void CallerInformationParametersShouldBeLastInvalidSyntax()
         {
-            Verifier.VerifyAnalyzer(@"TestCases\CallerInformationParametersShouldBeLastInvalidSyntax.cs",
+            // FIXME the update to .NET 5 changed this behavior from a raised issue to no issues raised
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/CallerInformationParametersShouldBeLastTest.cs:43]: I'm not sure if it's worth opening a ticket for this

File: scripts/build/build-utils.ps1 [hunk 3]
```diff
@@ -85,25 +85,6 @@ function New-NuGetPackages([string]$binPath) {
     }
 }
 
-function Restore-Packages (
```
Comment (andrei-epure-sonarsource) [inline scripts/build/build-utils.ps1:88]: before this commit, I had some weird local failures when running the `dev-build` script with `-restore`, errors like the one below
```
C:\Program Files\dotnet\sdk\5.0.100-rc.2.20479.15\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error NETSDK1005: Assets file 'C:\Workspace\sonar-dotnet\sonaranalyzer-dotnet\tests\SonarAnalyzer.UnitTest\obj\project.assets.json' does n't have a target for 'net48'. Ensure that restore has run and that you have included 'net48' in the TargetFrameworks for your project. [C:\Workspace\sonar-dotnet\sonaranalyzer-dotnet\tests\SonarAnalyzer.UnitTest\SonarAnalyzer.UnitTest.csproj]
```

File: azure-pipelines.yml [hunk 4]
```diff
@@ -60,6 +60,20 @@ stages:
           Write-Host "##vso[task.setvariable variable=SONAR_PROJECT_VERSION;]$sonarProjectVersion"
         displayName: 'Read Sonar project version from the versions.props file'
 
+      - powershell: |
+          cd sonaranalyzer-dotnet
+          $testProjFileName = "tests\SonarAnalyzer.UnitTest\SonarAnalyzer.UnitTest.csproj"
+
+          Write-Host "Remove .Net 5 target for UnitTest project"
```
Comment (costin-zaharia-sonarsource) [inline azure-pipelines.yml:67]: So we will later configure the environment. I think this is fine since it will be much easier after the official release. 

Is there an issue to not forget about this?

File: scripts/build/dev-build.ps1 [hunk 5]
```diff
@@ -62,12 +62,13 @@ try {
...
 
+    $buildTarget = "Rebuild"
     if ($restore) {
-        Restore-Packages $msbuildVersion $solutionName
+        $buildTarget = "Restore,Rebuild"
     }
 
     if ($build) {
-        Invoke-MSBuild $msbuildVersion $solutionName `
+        Invoke-MSBuild $msbuildVersion $solutionName /t:"$buildTarget" `
```
Comment (costin-zaharia-sonarsource) [inline scripts/build/dev-build.ps1:71]: I'm not sure if we have a scenario where we want to build without a restore. 

We might want to simplify this script and remove the parameter. We could also use `dotnet build` since it does a restore by default if needed.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/CallerInformationParametersShouldBeLastTest.cs (hunk 2 above)
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/CallerInformationParametersShouldBeLastTest.cs:43]: My concern is that we will never address this if we will not do it now; I would remove the comment.

Since the test is also checking that no errors are produced while analyzing invalid syntax I consider that it's still useful.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/ClassNameTest.cs [hunk 6]
```diff
@@ -44,8 +44,12 @@ public void ClassName_CSharp()
                     @"TestCases\ClassName.cs",
                     @"TestCases\ClassName.Partial.cs",
                 }, new CSharp.ClassAndMethodName(),
+#if NETFRAMEWORK
                 options: ParseOptionsHelper.FromCSharp8,
                 additionalReferences: NuGetMetadataReference.NETStandardV2_1_0);
+#else
+                options: ParseOptionsHelper.FromCSharp8);
+#endif
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/ClassNameTest.cs:52]: We can simplify this by adding `additionalReferences: NuGetMetadataReference.NETStandardV2_1_0` only for `NETFRAMEWORK`?

Since for named parameters the order does not matter we can have something like:
```
#if NETFRAMEWORK
	additionalReferences: NuGetMetadataReference.NETStandardV2_1_0,
#endif
	options: ParseOptionsHelper.FromCSharp8);
```

Another option would be to add a new method in `MetadataReferenceFacade` to get a reference to .Net Standard.

Edit: there are multiple occurrences of this. We might want to check where this applies and simplify them.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NativeMethodsShouldBeWrappedTest.cs [hunk 7]
```diff
@@ -33,7 +33,8 @@ public class NativeMethodsShouldBeWrappedTest
         public void NativeMethodsShouldBeWrapped()
         {
             Verifier.VerifyAnalyzer(@"TestCases\NativeMethodsShouldBeWrapped.cs",
-                new NativeMethodsShouldBeWrapped());
+                new NativeMethodsShouldBeWrapped(),
+                checkMode: CompilationErrorBehavior.Ignore);
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NativeMethodsShouldBeWrappedTest.cs:37]: why is this needed?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj [hunk 8]
```diff
@@ -1,12 +1,12 @@
-﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>preview</LangVersion>
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj:4]: Nice :)

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs [hunk 9]
```diff
@@ -64,8 +64,20 @@ public Compilation GetCompilation(ParseOptions parseOptions = null, CompilationO
         public Document FindDocument(string name) =>
             Project.Documents.Single(d => d.Name == name);
 
-        public ProjectBuilder AddReferences(IEnumerable<MetadataReference> references) =>
-            FromProject(Project.AddMetadataReferences(references ?? Enumerable.Empty<MetadataReference>()));
+        public ProjectBuilder AddReferences(IEnumerable<MetadataReference> references)
+        {
+            var existingReferences = Project.MetadataReferences;
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs:69]: Wouldn't be better to fix this in the test setup? This will allow us to be sloppy with metadata references.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs [hunk 10]
```diff
@@ -64,8 +64,20 @@ public Compilation GetCompilation(ParseOptions parseOptions = null, CompilationO
...
+            IEnumerable<MetadataReference> deduplicated;
+            if (references == null)
+            {
+                deduplicated = Enumerable.Empty<MetadataReference>();
+            }
+            else
+            {
+                deduplicated = references.Where(mr => !existingReferences.Contains(mr)).GroupBy(x => x.Display).Select(g => g.First()).ToHashSet();
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs:77]: Can't use distinct?

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs (hunk 9 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs:69]: there's a lot of tests to fix...

File: azure-pipelines.yml (hunk 4 above)
Comment (andrei-epure-sonarsource) [reply azure-pipelines.yml:67]: I opened https://github.com/SonarSource/sonar-dotnet/issues/3684

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NativeMethodsShouldBeWrappedTest.cs (hunk 7 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Rules/NativeMethodsShouldBeWrappedTest.cs:37]: lines 3-4 of the file:
```
// Invalid method, testing if IMethodSymbol.ContainingType returns null (it currently doesn't)
private void Do(int x) { } // Error [CS0116]
```

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs (hunk 9 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs:69]: After reverting, I got NN errors like below
```
  Message: 
    Test method SonarAnalyzer.UnitTest.Common.SecurityHotspotTest.SecurityHotspotRules_AreDisabledByDefault threw exception: 
    System.ArgumentException: Specified sequence has duplicate items
    Parameter name: metadataReferences[19]
```
image

**Comment by costin-zaharia-sonarsource** [comment]
> locally, I have one failing test
> 
> ```
>   Failed RestrictDeserializedTypesFormatters [68 ms]
>   Error Message:
>    Test method SonarAnalyzer.UnitTest.Rules.RestrictDeserializedTypesTest.RestrictDeserializedTypesFormatters threw exception:
> SonarAnalyzer.UnitTest.TestFramework.UnexpectedDiagnosticException: CSharp8: Unexpected primary issue on line 187, span (186,23)-(186,34) with message 'Restrict types of objects allowed to be deserialized.'.
> See output to see all actual diagnostics raised on the file
> ```
> 
> It's good to keep in mind...

This fails also on the build pipeline. You can find the logs [here](https://sonarsource.visualstudio.com/DotNetTeam%20Project/_build/results?buildId=23272&view=logs&j=23c99647-d7e1-5b83-b5fb-f7e97f8a9ce7&t=0075a4ce-4d44-5a4c-36ef-97ab11b0917b&l=6858).

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs (hunk 9 above)
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs:69]: I see... I think the effort is to big and it's not worth it. Maybe we will take a second look after we change the NuGet api.

**Comment by andrei-epure-sonarsource** [comment]
>It seems that we need to update the script to fail the build in this case.

I'll open a different PR to fix this - see https://github.com/SonarSource/sonar-dotnet/issues/3686

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs (hunk 9 above)
Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs:69]: I put a comment in #3674 to keep it in mind https://github.com/SonarSource/sonar-dotnet/issues/3674#issuecomment-717857880

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM! 🚀
